### DISCUSSION TITLE
Updating context within an action or service

### DISCUSSION THREAD

**Discussion by BMR11** [statelyai/xstate#3410] [Q&A]
First, a big thank you for this awesome state management library.

I learnt like to update the context there is only one way and that is to use assign action.
My case is something like 

actions:["doSomeStuffs"]

and on doSomeStuffs, I want to do something like
doSomeStuffs = (context, event) => {

//update context 

await getSomeData()

//update context 

await getSomeData2()

//update context

}

Here, I am not sure how can I update context. assign is an action so I can not directly call. I can do like to split doSomeStuffs into individual actions where I can use assign actions as per sequence I need. But I think there should be a better way of doing this.

Thank you

**Comment by Andarist** [comment]
You could wrap this in an invoked actor and send events to the parent that in turn would execute `assign` actions there.

**Comment by BMR11** [comment]
Thank you Andarist for reply.
I think I am using similar thing to make it working but this would break big operation into separate calls unnecessarily. 

Can you provide some sample code for my case above so I can get this correctly?

Thank you

**Comment by james-yap** [comment]
I have recently stumbled upon a similar issue:- I was trying to update the machine context from within an [invoked promise actor](https://stately.ai/docs/promise-actors).

I was aware that actions could be used to update context (see [Assign action](https://stately.ai/docs/actions#assign-action)), but actions run outside of a promise and do not guarantee order. However, I have figured out the way to do it which I believe is the way the maintainer intended.

Here is the [invoked promise actor](https://stately.ai/docs/promise-actors):

```typescript
submitJob: fromPromise(async ({ input }: { input: { name: string } }) => {
  console.log('Starting submitJob', input);
  return { jobuid: '123' };
}),
```

And this is the state that invokes this actor:

```typescript
SubmitJob: {
  invoke: {
    src: 'submitJob',
    input: ({ context }) => ({
      name: context.job.name
    }),
    onDone: {
      target: 'WaitForCompletion',
      actions: assign({
        jobuid: ({ event }) => event.output.jobuid
      })
    }
  }
},
```

The secret sauce is to use the `actions: assign()` field in the `onDone` property of the invoke statement. This is an action that will only run after the promise actor is completed. And since we can send events to the action from the promise actor, we can use the action to update the context with information from this event. For more information on `onDone`, check out this page from the docs [here](https://stately.ai/docs/invoke#ondone).

An excellent example that I drew from:
https://github.com/statelyai/xstate/blob/main/examples/workflow-monitor-job/main.ts
